rules/opentitan.bzl

```diff
@@ -0,0 +1,224 @@
+        outputs.append(scrambled)
+        ctx.actions.run(
+            outputs = [scrambled],
+            inputs = [
+                src,
+                ctx.files._tool[0],
+                ctx.files._config[0],
+            ],
+            arguments = [
+                ctx.files._config[0].path,
+                src.path,
+                scrambled.path,
+            ],
+            executable = ctx.files._tool[0].path,
+        )
+        return [DefaultInfo(files = depset(outputs), data_runfiles = ctx.runfiles(files = outputs))]
+
+elf_to_scrambled = rule(
+    implementation = _elf_to_scrambled,
+    cfg = _platforms_transition,
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
+        "platform": attr.string(default = OPENTITAN_PLATFORM),
+        "_tool": attr.label(default = "//hw/ip/rom_ctrl/util:scramble_image.py", allow_files = True),
+        "_config": attr.label(default = "//hw/top_earlgrey:data/autogen/top_earlgrey.gen.hjson", allow_files = True),
+        "_allowlist_function_transition": attr.label(
+            default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+        ),
+    },
+)
+
+def opentitan_binary(
+        name,
+        platform = OPENTITAN_PLATFORM,
+        per_device_deps = {
+            "verilator": ["//sw/device/lib/arch:sim_verilator"],
+            "dv": ["//sw/device/lib/arch:sim_dv"],
+            "fpga_nexysvideo": ["//sw/device/lib/arch:fpga_nexysvideo"],
+            "cw310": ["//sw/device/lib/arch:fpga_cw310"],
+        },
+        output_bin = True,
+        output_disassembly = True,
+        output_scrambled = False,
+        **kwargs):
+    """A helper macro for generating OpenTitan binary artifacts.
+    This macro is mostly a wrapper around cc_binary, but creates artifacts
+    for each of the keys in `per_device_deps`.  The actual artifacts
+    created are an ELF file, a BIN file, the disassembly and the scrambled
+    ROM image.  Each of these output targets performs a bazel transition to
+    the RV32I toolchain to build the target under the correct compiler.
+    Args:
+      @param name: The name of this rule.
+      @param platform: The target platform for the artifacts.
+      @param per_device_deps: The deps for each of the execution environments.
+      @param output_bin: Whether or not to emit a BIN file.
+      @param output_disassembly: Whether or not to emit a disassembly file.
+      @param output_scrambled: Whether or not to emit a SCR file.
+      @param **kwargs: Arguments to forward to `cc_binary`.
+    Emits rules:
+      For each device in per_device_deps entry:
+        cc_binary     named: name_device
+        obj_transform named: name_device_elf
+        optionally:
+          obj_transform       named: name_device_bin
+          elf_to_dissassembly named: name_device_dis
+          elf_to_scrambled    named: name_device_scr
+      filegroup named: name
+          with all the srcs used
```

Comment:
nit: the filegroup rule names all of the previously generated rules (everything under "for each") as members of the filegroup.